### PR TITLE
Fixed #24793 -- Unified temporal subtraction support.

### PR DESCRIPTION
**Pull request by charettes**

### CODE REVIEW

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -448,6 +452,33 @@ def as_sql(self, compiler, connection):
...
+
+    def as_sql(self, compiler, connection):
+        expressions = []
+        expression_params = []
+        sql, params = compiler.compile(self.lhs, connection)
+        expressions.append(sql)
+        expression_params.extend(params)
+        sql, params = compiler.compile(self.rhs, connection)
+        expressions.append(sql)
+        expression_params.extend(params)
```
Comment (charettes) [inline django/db/models/expressions.py]: The following should be moved to the backends.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Comment by claudep** [comment]
Impressive!

**Comment by timgraham** [comment]
I guess this should be documented in at least the "Database backend API" section of the release notes. Maybe "Minor features" too? (not sure if it's more of a bug fix)

**Comment by charettes** [comment]
Now the question is whether we want to support different temporal subtraction, e.g. `DateTimeField` - `TimeField`. I can see value in implementing it for the `DateField` and `DateTimeField` but I'd like to hear the ORM guys about it,

Here's the different signature of the `-` operation on PostgreSQL for reference:

LHS type | RHS type | Return type
------------ | ------------- | -------------
date | date | integer
date | time | timestamp
date | timestamp | interval
time | date | _undefined_
time | time | interval
time | timestamp | _undefined_
timestamp |date | interval
timestamp | time | timestamp
timestamp | timestamp | interval

**Comment by timgraham** [comment]
Is that within the scope of this ticket? I guess you should ask on the mailing list.

**Comment by Ian-Foote** [comment]
Python's datetime module doesn't support subtraction between `datetime`, `date` and `time`, so I think it would be surprising if Django fields did support it.

**Comment by charettes** [comment]
Thanks for your feedback Ian, it makes sense not to support it in this case.

**Pull request merged**